### ISSUE TITLE
`lower_bound_async`

### ISSUE THREAD

**Issue by BenBrock** [oneapi-src/oneDPL#930]
I need an asynchronous version of `lower_bound` in order to implement a multi-GPU sample sort algorithm.

It looks like it should be fairly straightforward to implement `lower_bound_async`, since the [current implementation of `lower_bound`](https://github.com/oneapi-src/oneDPL/blob/main/include/oneapi/dpl/internal/binary_search_impl.h#L111) just calls `parallel_for`.

If I were to submit a PR, what would be the correct file to add an implementation of `lower_bound_async`?  `async_impl_hetero.h`? `glue_async_impl.h`?  Please advise and I'll take a stab at it.

**Comment by reble** [comment]
Is this still relevant? This recent PR might be worth to look at for you: 
https://github.com/oneapi-src/oneDPL/pull/903/files

In the `glue*.h` file you would map the algorithm including different signatures to a specific pattern. This is aligned with the common infrastructure for algorithm. You can find examples here:
https://github.com/oneapi-src/oneDPL/blob/main/include/oneapi/dpl/pstl/hetero/algorithm_impl_hetero.h

We would also need some basic tests.

**Comment by BenBrock** [comment]
> Is this still relevant? This recent PR might be worth to look at for you:

Yes, I think this is still relevant---that PR modifies async `sort`, but what I need is async `lower_bound` (as part of a distributed sort).

I'm a bit lost with the patterns---is the idea that in the async implementation, [instead of calling `parallel_for` like is done here](https://github.com/oneapi-src/oneDPL/blob/main/include/oneapi/dpl/internal/binary_search_impl.h#L131), I'd call something like [`__patern_walk`](https://github.com/oneapi-src/oneDPL/blob/main/include/oneapi/dpl/pstl/hetero/algorithm_impl_hetero.h#L44)?
